### ISSUE TITLE
Compiler warnings (mingw)

### ISSUE THREAD

**Issue by FreeBASIC-programmer** [phorward/phorward#1]
When compiling the code (version 0.20.0) using msys/mingw(32 bit) on windows 7 (64 bit pro) I get the following warnings:
```
string/convert.c: In function `pdbl_to_wcs':
string/convert.c:164: warning: passing arg 2 of `swprintf' makes pointer from integer without a cast

string/string.c: In function `pwcscatchar':
string/string.c:884: warning: passing arg 2 of `swprintf' makes pointer from integer without a cast

string/utf8.c:33: warning: large integer implicitly truncated to unsigned type
string/utf8.c:34: warning: large integer implicitly truncated to unsigned type
string/utf8.c:34: warning: large integer implicitly truncated to unsigned type
string/utf8.c:35: warning: large integer implicitly truncated to unsigned type
string/utf8.c: In function `u8_toutf8':
string/utf8.c:235: warning: comparison is always true due to limited range of data type
string/utf8.c:242: warning: comparison is always true due to limited range of data type
string/utf8.c: In function `u8_wc_toutf8':
string/utf8.c:268: warning: comparison is always true due to limited range of data type
string/utf8.c:274: warning: comparison is always true due to limited range of data type
string/utf8.c: In function `u8_escape_wchar':
string/utf8.c:468: warning: comparison is always false due to limited range of data type
string/utf8.c:470: warning: comparison is always true due to limited range of data type

libtool: warning: undefined symbols not allowed in i686-pc-mingw32 shared libraries; building static only
```
The swprintf problem can be resolved by adding the following snippet to both string.c and convert.c
```
#if _WIN32
#define swprintf _snwprintf
#endif
```
The utf8.c related warnings have to do with the size of wchar_t on windows (unsigned 16 bit integer). On Linux wchar_t is defined as a 32 bit integer which means that utf8.c will most likely compile without a problem on Linux. 

The author you got the utf8 - code from has this to say about his utf8 library:

> I now use and recommend  [utf8proc](https://github.com/JuliaLang/utf8proc ) instead of this library.

**Comment by phorward** [comment]
Hi @FreeBASIC-programmer, thank your for submitting this issue.

After setting up a test environment (Win 10, latest MSYS), I already failed on compiling any file because wctype.h is not available in the system. Did you got the same problem before? 

Removing the include gets me to the same result as you reported, as it can be seen in the attached screenshot. ![virtualbox_msedge - win10_26_11_2017_22_43_43](https://user-images.githubusercontent.com/16870072/33244653-b8748b3a-d2fb-11e7-8129-49d924233f09.png)

The binaries can't be linked due missing unicode functions.
![virtualbox_msedge - win10_26_11_2017_22_46_33](https://user-images.githubusercontent.com/16870072/33244660-ca055b54-d2fb-11e7-9a40-8fd1fbb15e90.png)

It seems that the UNICODE/UTF8 detection does not work properly. Some more tests need to be done.

-- Jan

**Comment by FreeBASIC-programmer** [comment]
I did not get a problem when compiling the code other than the warnings as mentioned in my previous post (mingw created a static library called libphorward.a).

I  am using the following (very old) version of mingw (which includes wctypes.h)
```
$ gcc --version
gcc.exe (GCC) 3.4.5 (mingw-vista special r3)
Copyright (C) 2004 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```
The msys version I am using is msys 1.0.11.

**Comment by phorward** [comment]
Hey, I did some restructuring of the dependencies in configure.ac, and successfully tested it with the latest MSYS on a 32-Bit Windows 8.1. The mingw-gcc that I use is 6.3.0.

Can you please have a try on 9a236f323bd74519c204de3c044324b9d9df5536?

**Comment by FreeBASIC-programmer** [comment]
I compiled 9a236f3 and got the same warnings.

**Comment by phorward** [comment]
OK now I re-tested it with the same version that you have and was able to reproduce your warnings. I now cleared the warnings by doing some pre-processor switching of code, but I am sure that there are still many bugs in the related functions on platforms with a 16-Bit wchar_t.

Maybe, the wide-character support will be removed in a future version due lacking of usage and portability. I think, writing tests would be a good option, but then for all cases and platforms.

**Comment by phorward** [comment]
The issue should be fixed by commit 3dd05c1defd26a2b0ca289bc1b7c953d880b53ad.
